Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,245 @@
+            auto baseURLPath = processBaseURLString(baseURL.path(), type);
+            size_t slashIndex = baseURLPath.reverseFind('/');
+
+            if (slashIndex != notFound)
+                result.pathname = makeString(StringView { baseURLPath }.left(slashIndex + 1), result.pathname);
+
+            auto pathResult = canonicalizePath(result.pathname, baseURL.protocol(), type);
+
+            if (pathResult.hasException())
+                return pathResult.releaseException();
+
+            result.pathname = pathResult.releaseReturnValue();
+        }
+    }
+
+    if (!init.search.isNull()) {
+        auto queryResult = canonicalizeSearch(init.search, type);
+
+        if (queryResult.hasException())
+            return queryResult.releaseException();
+
+        result.search = queryResult.releaseReturnValue();
+    }
+
+    if (!init.hash.isNull()) {
+        auto fragmentResult = canonicalizeHash(init.hash, type);
+
+        if (fragmentResult.hasException())
+            return fragmentResult.releaseException();
+
+        result.hash = fragmentResult.releaseReturnValue();
+    }
+
+    return result;
+}
+
+ExceptionOr<Ref<URLPattern>> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
 {
     UNUSED_PARAM(baseURL);
 
-    return adoptRef(*new URLPattern());
+    return Exception { ExceptionCode::NotSupportedError, "Not implemented."_s };
 }
 
-Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&&, URLPatternOptions&&)
+ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&& input, URLPatternOptions&&)
 {
-    return adoptRef(*new URLPattern());
+    URLPatternInit init;
+    if (!input)
+        return Exception { ExceptionCode::NotSupportedError, "Not implemented."_s };
+    if (std::holds_alternative<String>(*input) && !std::get<String>(*input).isNull())
+        return Exception { ExceptionCode::NotSupportedError, "Not implemented."_s };
+    if (std::holds_alternative<URLPatternInit>(*input))
+        init = std::get<URLPatternInit>(*input);
+
+    auto maybeProcessedInit = processInit(WTFMove(init), BaseURLStringType::Pattern);
+
+    if (maybeProcessedInit.hasException())
+        return maybeProcessedInit.releaseException();
+
+    auto processedInit = maybeProcessedInit.releaseReturnValue();
+
+    // FIXME: Implement compiling the component for each URLPattern member field to replace this current placeholder code which reroutes processed URLPatternInit object to URLPattern result.
+    return adoptRef(*new URLPattern(WTFMove(processedInit)));
 }
 
-URLPattern::URLPattern() = default;
+URLPattern::URLPattern(URLPatternInit&& initInput)
+{
+    m_protocol = initInput.protocol;
+    m_username = initInput.username;
+    m_password = initInput.password;
+    m_hostname = initInput.hostname;
+    m_port = initInput.port;
+    m_pathname = initInput.pathname;
+    m_search = initInput.search;
+    m_hash = initInput.hash;
+}
 
+URLPattern::URLPattern() = default;
```

Comment:
Do we still need this one?

Reply:
The Blink implementation has it so I'd like to keep it around.